pyccel/ast/builtin_objects/list_functions.py

```diff
@@ -0,0 +1,77 @@
+# coding: utf-8
+#------------------------------------------------------------------------------------------#
+# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
+# go to https://github.com/pyccel/pyccel/blob/master/LICENSE for full license details.     #
+#------------------------------------------------------------------------------------------#
+"""
+The List container has a number of built-in methods that are 
+always available.
+
+This module contains objects which describe these methods within Pyccel's AST.
+"""
+
+from pyccel.ast.datatypes import NativeVoid, NativeGeneric, NativeHomogeneousList
+from pyccel.ast.internals import PyccelInternalFunction
+
+
+class ListAppend(PyccelInternalFunction):
+    """
+    Represents a call to the .append() method.
+
+    Represents a call to the .append() method of an object with a list type,
+    which adds an element to the end of the list. This method returns `None`.
+    The append method is called as follows:
+
+    >>> a = [1]
+    >>> a.append(2)
+    >>> print(a)
+    [1, 2]
+
+    Parameters
+    ----------
+    lst_bound_arg : TypedAstNode
+        The variable representing the list.
+    
+    new_elem : TypedAstNode
+        The argument passed to append() method.
+    """
+    __slots__ = ("_lst_bound_arg", "_append_arg")
+    _dtype = NativeVoid()
+    _shape = None
+    _order = None
+    _rank = 0
+    _precision = -1
+    _class_type = NativeHomogeneousList()
+    name = 'append'
+
+    def __init__(self, lst_bound_arg, new_elem) -> None:
+        conditions = (
+            new_elem.dtype is not NativeGeneric() and
+            lst_bound_arg.dtype == new_elem.dtype and
+            lst_bound_arg.precision == new_elem.precision and
+            lst_bound_arg.rank - 1 == new_elem.rank
+        )
+        is_homogeneous = lst_bound_arg.dtype is not NativeGeneric() and conditions
+        if not is_homogeneous:
+            raise TypeError("Expecting an argument of the same type as the elements of the list")
+        self._lst_bound_arg = lst_bound_arg.name
```

Comment:
There seems to be a problem here. Your docstring says that you take a `TypedAstNode` but that type doesn't have a `name` property. How can you be sure that you can access this property?

Reply:
there's two cases that i have checked:
```python
a = [4]
b = 3
a.append(b)
a.append(5)
```
`name` property in the first statement is `b`,while in the second statement it would be something like `dummy_000`
in which case do you think the passed variable wouldn't have a `name` attribute?

Reply:
Interesting, so `lst_bound_arg` doesn't represent the list on which we are operating? I expected `name` to be `a` not `b`.
If the name is `dummy_000` then the semantic stage must be creating a temporary variable to store the argument. I didn't think it did this for literals. However if it does then that is fine, but you are not handling `TypedAstNode`, you are handling `Variable` so your docstring needs correcting

Reply:
I have managed to think of an example that should break:
```python
class A:
    def __init__(self):
         self.my_list : list[int] = []

    def get_list(self):
        return self.my_list
        
if __name__ == '__main__':
    my_a = A()
    my_a.get_list().append(4)
```

Reply:
yes i see your point, do you have any suggestions to fix this?

Reply:
Don't save the name, just save a `TypedAstNode` and let the printer decide how to print it

Reply:
fixed